### ISSUE TITLE
UnitOfWork.computeChangeSets() expensive on large queries

### ISSUE THREAD

**Issue by parisholley** [mikro-orm/mikro-orm#2379]
**Describe the bug**

This snippet is mimicking an endpoint which is doing a large query, calculating some aggregates then spitting out a stats object, without making any model modifications. My app is setup to auto-flush request forks at the end of each request.

```typescript
const jobs = await em.find(MyModel, {...}); // returns 3K rows

await em.flush(); // effectively hangs the process
```

Doing some debugging, it looks like there is a very expensive call happening on all 3K rows and even worse, it loops through each entity identity map, which has a ton of uninitialized References.

https://github.com/mikro-orm/mikro-orm/blob/f09a951d2f4725e631a66e7219483efb1e4437c8/packages/core/src/unit-of-work/UnitOfWork.ts#L286-L291

If I comment out line 289, the response returns in a reasonable amount of time. I believe the problem is there is some recursion going on with the bi-directional mappings, and we are using a `new WeakSet<AnyEntity>()` on every iteration vs creating a single instance.

**Comment by B4nan** [comment]
I woudn't suggesting flushing each request blindly, there will always be a performance penalty. We need to go thru all entities inside identity map, and we need to cascade, otherwise things simply won't work as they should (e.g. persistence by reachability). I don't think it could be caused by the `new WeakSet()` call itself (but you could verify that easily, if you see this locally). References that are not initialized should be simply skipped, that shouldn't matter either.

Please provide full reproduction for this if you want me to look into this.

**Comment by B4nan** [comment]
I will need a reproduction for this. Tried it this way but it looks fast enough to me (postgres):

```ts
for (let i = 1; i <= 3000; i++) {
  const author = new Author2(`Jon Snow ${i}`, `snow-${i}@wall.st`);
  author.books.add(new Book2('b', author), new Book2('b', author));
  orm.em.persist(author);
}

let now = Date.now();
await orm.em.flush();
console.log(Date.now() - now); // insert 3k authors + 6k books = 650ms
orm.em.clear();

const ret = await orm.em.find(Author2, {}, { populate: ['books'] });
now = Date.now();
await orm.em.flush(); // check all those 9k entities for changes = 100ms
console.log(Date.now() - now);
```

(and I am running this via ts-jest, so far from production performance)

**Comment by parisholley** [comment]
i'll be able to put a reproduction together later this week. 

in the interim, i think you'd have a better change to reproduce if you have at least 3 models, which have bi-directional 1:M and M:1 between them, so:

Author <> Books
Author <> Genre (eg: "Primary Genre")
Book <> Genre  

the issue came down to recursion between models, where it would cascade to a collection, and the collection would include a model that was already hit in another model/collection.

**Comment by B4nan** [comment]
Any updates on this? Tried once again, it still looks ok to me, at least in master/v5.

```ts
test('perf: complex relations', async () => {
  let now = Date.now();

  for (let i = 1; i <= 3000; i++) {
    const author = new Author2(`Jon Snow ${i}`, `snow-${i}@wall.st`);
    const b1 = new Book2(`b ${i} 1/3`, author);
    const b2 = new Book2(`b ${i} 2/3`, author);
    const b3 = new Book2(`b ${i} 3/3`, author);

    for (let j = 1; j <= 20; j++) {
      b1.tags.add(new BookTag2(`bt ${i} 1/3 ${j}/20`));
      b2.tags.add(new BookTag2(`bt ${i} 2/3 ${j}/20`));
      b3.tags.add(new BookTag2(`bt ${i} 3/3 ${j}/20`));
    }

    author.books.add(b1, b2, b3);
    orm.em.persist(author);
  }

  // creates 3k authors, 9k books and 180k tags in memory
  console.log(Date.now() - now); // ~3s

  now = Date.now();
  await orm.em.flush();

  // flush those 39k entities to database
  console.log(Date.now() - now); // ~17s

  orm.em.clear();
  now = Date.now();
  // load those 39k entities to memory
  const a = await orm.em.find(Author2, {}, { populate: ['books.tags'] });
  console.log(Date.now() - now); // ~9s
  now = Date.now();

  // compute changesets on them
  await orm.em.flush();
  console.log(Date.now() - now); // ~800ms
});
```

This is with postgres. Calling flush on just the 3k authors, without populating the relations, takes just 50ms. Also tried to put there some m:1 relations (as collections should have no effect unless being populated), still very similar numbers.

The numbers could be better for sure, but given this test is working with almost 200k entities, it seems quite fine to me, most of the time is spend on actual database queries. The initial 3s is quite a lot for constructing the entity graph, but this is caused by the propagation, and again, too many entities in the memory.

**Comment by B4nan** [comment]
Still nothing?

Can't say I am happy having open issue for a month with no actual reproduction.

**Comment by parisholley** [comment]
```typescript
import {
  BigIntType,
  Collection,
  Entity, Enum,
  IdentifiedReference,
  ManyToMany,
  ManyToOne,
  MikroORM, OneToMany,
  PrimaryKey,
  Property
} from '@mikro-orm/core';
import type { SqliteDriver } from '@mikro-orm/sqlite';

@Entity({ tableName: 'member_relationship'})
export class VendorBuyerRelationshipMikro {
  @PrimaryKey({ type: BigIntType, fieldName: 'id', setter: true })
  id!: string;

  @Property({ fieldName: 'created', onCreate: () => new Date() })
  created!: Date;

  @ManyToOne(() => MemberMikro, { fieldName: 'activeId', wrappedReference: true })
  buyer!: IdentifiedReference<MemberMikro>;

  @ManyToOne(() => MemberMikro, { fieldName: 'passiveId', wrappedReference: true })
  vendor!: IdentifiedReference<MemberMikro>;

  @OneToMany(() => OrderMikro, (o) => o.buyerRel)
  orders = new Collection<OrderMikro>(this);
}

@Entity({ tableName: 'member' })
export class MemberMikro {
  @PrimaryKey({ type: BigIntType, fieldName: 'id', setter: true })
  id!: string;

  @Property({ fieldName: 'created', onCreate: () => new Date() })
  created!: Date;

  @OneToMany(() => MemberMikro, (member) => member.parent)
  children = new Collection<MemberMikro>(this);

  @OneToMany(() => VendorBuyerRelationshipMikro, (rel) => rel.vendor)
  buyers = new Collection<VendorBuyerRelationshipMikro>(this);

  @OneToMany(() => OrderMikro, (order) => order.vendor)
  orders = new Collection<OrderMikro>(this);

  @ManyToOne(() => MemberMikro, { fieldName: 'parentId', wrappedReference: true })
  parent?: IdentifiedReference<MemberMikro>;
}

@Entity({ tableName: 'job' })
export class JobMikro {
  @PrimaryKey({ type: BigIntType, fieldName: 'id', setter: true })
  id!: string;

  @ManyToOne(() => MemberMikro, { fieldName: 'memberId', wrappedReference: true })
  member?: IdentifiedReference<MemberMikro>;

  @ManyToOne(() => OrderMikro, { fieldName: 'orderId', wrappedReference: true })
  order?: IdentifiedReference<OrderMikro>;

  @OneToMany(() => JobMikro, (job) => job.parent)
  children = new Collection<JobMikro>(this);

  @ManyToOne(() => JobMikro, { fieldName: 'parentId', wrappedReference: true })
  parent!: IdentifiedReference<JobMikro>;

  @Property()
  rejected = false;

  @ManyToOne(() => VendorBuyerRelationshipMikro, { fieldName: 'buyerRelId', wrappedReference: true })
  buyer?: IdentifiedReference<VendorBuyerRelationshipMikro>;

  @ManyToOne(() => JobMikro, { fieldName: 'delegateId', wrappedReference: true })
  delegate?: IdentifiedReference<JobMikro>;

  @ManyToOne(() => MemberMikro, { fieldName: 'assigneeId', wrappedReference: true })
  assignee?: IdentifiedReference<MemberMikro>;
}

@Entity({ tableName: 'order' })
export class OrderMikro {
  @PrimaryKey({ type: BigIntType, fieldName: 'id', setter: true })
  id!: string;

  @Property({ fieldName: 'created', onCreate: () => new Date() })
  created!: Date;

  @ManyToOne(() => VendorBuyerRelationshipMikro, { fieldName: 'buyerRelId', wrappedReference: true })
  buyerRel!: IdentifiedReference<VendorBuyerRelationshipMikro>;

  @OneToMany(() => JobMikro, (job) => job.order)
  jobs = new Collection<JobMikro>(this);

  @ManyToOne(() => MemberMikro, { fieldName: 'vendorId', wrappedReference: true })
  vendor!: IdentifiedReference<MemberMikro>;
}

describe('GH issue 2379', () => {
  jest.setTimeout(120000);

  let orm: MikroORM<SqliteDriver>;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [OrderMikro, JobMikro, VendorBuyerRelationshipMikro,MemberMikro],
      dbName: ':memory:',
      type: 'sqlite',
    });
    await orm.getSchemaGenerator().createSchema();
  });

  afterAll(() => orm.close(true));

  test('test cpu issues', async () => {
    const rels = [];

    const vendor = new MemberMikro();

    orm.em.persist(vendor);

    for (let i = 1; i <= 1000; i++) {
      const buyer = new MemberMikro();

      orm.em.persist(buyer);

      const rel = orm.em.create(VendorBuyerRelationshipMikro, {
        vendor,
        buyer
      });

      rels.push(rel);
    }

    for (let i = 1; i <= 3000; i++) {
      const order = orm.em.create(OrderMikro, {
        buyerRel: rels[i % 1000],
      });

      orm.em.persist(order);

      orm.em.create(JobMikro, {
        order,
      });
    }

    // this causes jest timeout error, even when bumped up to 120s
    await orm.em.flush();
    orm.em.clear();

    const jobs = await orm.em.find(JobMikro, { }, { populate: ['order'] });

    await orm.em.flush();
  });

});
```

this should be able to reproduce it, however i think the same issue exists on insert because it is taking too long on the initial flush. obviously in a real environment, we wouldn't be inserting this much upfront

**Comment by parisholley** [comment]
the solution i've been using for fixing the last query/flush, is to update `computeChangeSets()`, so that it shares the same visited weakset vs creating a new one on every loop

```typescript
  computeChangeSets(): void {
    const visited = new WeakSet();
    this.changeSets.clear();

    for (const entity of this.persistStack) {
      this.cascade(entity, Cascade.PERSIST, new WeakSet<AnyEntity>(), { checkRemoveStack: true });
    }

    for (const entity of this.identityMap) {
      if (!this.removeStack.has(entity) && !this.orphanRemoveStack.has(entity)) {
        this.persistStack.add(entity);
        this.cascade(entity, Cascade.PERSIST, visited, { checkRemoveStack: true });
      }
    }
```

**Comment by B4nan** [comment]
Perfect, thanks!

> the solution i've been using for fixing the last query/flush, is to update computeChangeSets(), so that it shares the same visited weakset vs creating a new one on every loop

I was actually thinking about that as well.

**Comment by B4nan** [comment]
So in v5/master, if I use the shared set for `visited`, it looks all fine, including the final flush. 

On my m1 the initial flush takes ~320ms, find takes ~200ms and the final flush takes ~265ms. with v4 there is also issue with calling `em.persist()`, so it hangs in the initial phase before first flush, in v5 cascading is delayed till flush which fixes this (84b5d1b1b867de7cfa8e1349327dc96b7250d4eb), so I can backport that too.

**Comment by B4nan** [comment]
So I cherry picked the existing fix and added the one with shared `visited` set to 4.x branch, can you please test it? It is available as `^4.5.10-dev.29`

**Comment by sergeyampo** [comment]
@parisholley Thanks to you, it solves a big performance problem in our case which was hard to reproduce.
